### PR TITLE
PARQUET-2347: Add interface layer between Parquet and Hadoop Configuration

### PR DESCRIPTION
**Pull request by amousavigourabi**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-1234: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
       Additional parameters to run the read/write tests using the new interface-using methods.

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

---

jacicmp exclusions have been added for the following classes: `org.apache.parquet.hadoop.CodecFactory`, `org.apache.parquet.hadoop.ParquetReader`. When these exclusions are removed, the following incompatibilities are detected:
```
There is at least one incompatibility: org.apache.parquet.hadoop.CodecFactory.configuration:FIELD_TYPE_CHANGED,org.apache.parquet.hadoop.ParquetReader$Builder.conf:FIELD_TYPE_CHANGED
```

This PR is part of an effort that has been [discussed on the dev mailing list](https://lists.apache.org/thread/4wl0l3d9dkpx4w69jx3rwnjk034dtqr8).

### CODE REVIEW

**Comment by wgtmac** [comment]
Thanks for working on this! I will take a look by the end of this week.

File: parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java
Comment (wgtmac) [inline parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java:412]: It seems that original ParquetReadOptions does not require any Configuration parameter. Should we avoid adding this by using ParquetConfiguration internally in the HadoopCodecs?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (wgtmac) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:176]: CMIW, we do not need to address the Hadoop dependency in the parquet-hadoop module, right?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (wgtmac) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:266]: If my previous statement is correct, then perhaps we do not need change in this line?

File: pom.xml
Comment (wgtmac) [inline pom.xml:550]: These lines worth adding some comments.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java
Comment (wgtmac) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java:438]: Does it mean that we still cannot get rid of the hadoop dependency? Anyway we have to depend on Configuration class here.

File: parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftReadSupport.java
Comment (wgtmac) [inline parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftReadSupport.java:311]: nit: could we give T1 and T2 better names?

**Review by wgtmac — COMMENTED** [review]
Thanks for doing this! I have left some inline comments. 

BTW, I am also curious of the following things.
- It seems that we still have some compatibility issues. Can you confirm? If yes, could you please write them out explicitly?
- Is there any follow-up work item to do? Would be good if we can know the whole picture in advance.
- Is it possible to add a simple test case to prove that a simple writer and reader roundtrip can work successfully without Hadoop dependency?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java
Comment (amousavigourabi) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:265]: Here, the code tries to load a Hadoop codec from its classloader. For this, the Hadoop classloader is needed. This is not available by just using getClassLoader() on any ParquetConfiguration.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java
Comment (amousavigourabi) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:267]: This PR removes _part_ of the Hadoop dependency. For at least the codecs we still rely on them OOTB (see PARQUET-2353). AFAIK, using uncompressed codec should not lead to the Hadoop classes being loaded (which would require the runtime dependency).

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java
Comment (amousavigourabi) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java:438]: For now this seems necessary to me in order not to introduce any breaking changes. AFAIK clients _could_ make suppClass be their own implementation, which under Parquet prior to this change would involve invoking setConf(Configuration) if that class were Hadoop Configurable (which is what Hadoop's ReflectionUtils.newInstance(Class, Configuration) also does under the hood).

File: parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/AbstractThriftWriteSupport.java
Comment (amousavigourabi) [reply parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/AbstractThriftWriteSupport.java:60]: `Class` is a generic type, raw use of which is discouraged. This is why I chose to give the generic `Class<?>` with a wildcard here.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ConfigurationUtil.java
Comment (amousavigourabi) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ConfigurationUtil.java:59]: When using a HadoopParquetConfiguration, the user did not yet decouple from Hadoop as it is just a wrapper for Configuration. When the user wants to decouple from Hadoop, they can implement their own ParquetConfiguration, which does not rely on Hadoop's Configuration (or a simple implementation is added afterwards, this PR was getting a bit large for that already). There is still some code right now, mainly around the codecs which _needs_ a Hadoop Configuration. It is therefore important that while we're still removing these last references to Hadoop, we can get such an instance from a ParquetConfiguration, in order not to break anything.

File: parquet-pig/src/main/java/org/apache/parquet/pig/TupleReadSupport.java
Comment (amousavigourabi) [reply parquet-pig/src/main/java/org/apache/parquet/pig/TupleReadSupport.java]: Agreed

File: parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftReadSupport.java
Comment (amousavigourabi) [reply parquet-thrift/src/main/java/org/apache/parquet/hadoop/thrift/ThriftReadSupport.java]: Done

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java
Comment (amousavigourabi) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordReader.java:176]: The Hadoop specific stuff I agree that it would be a bit silly. However, this class `InternalParquetRecordReader` is part of the read/write API for which we're trying to address it. That the read/write API is part of parquet-hadoop is somewhat unfortunate, but changing it at this point would break stuff.

File: parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java
Comment (amousavigourabi) [reply parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java:412]: We still need a way to get the ParquetConfiguration to the HadoopCodecs, for this I used a field in ParquetReadOptions. To this effect, I would also like to start a conversation in the future about the options classes, as it seems to me we can replace them with the more versatile ParquetConfiguration in most, if not all places. This as I do think this current flow of passing the ParquetConfiguration through the ParquetReadOptions is a bit hacky.

**Comment by amousavigourabi** [comment]
@wgtmac thanks a lot for the review! It's quite a big one so I really appreciate you took the time for it.

To address your concerns:

> * It seems that we still have some compatibility issues. Can you confirm? If yes, could you please write them out explicitly?

Correct, japicmp comes up with two incompatibilities. These are in the classes CodecFactory and ParquetReader. The incompatibilities it points to are both changes of private and protected field types from Configuration to ParquetConfiguration. These changes are strictly necessary for the effort to unhadoop the read/write API.

> * Is there any follow-up work item to do? Would be good if we can know the whole picture in advance.

After this, the following steps that will need to be taken are: 1. the creation of a simple unhadooped implementation of the ParquetConfiguration interface, and 2. adding a simple way for users to avoid the Hadoop codecs, as the OOTB implementations of everything still rely very heavily on Hadoop classes. These changes should allow users to drop the Hadoop runtime dependency. The Hadoop client API dependency will still be necessary.

> * Is it possible to add a simple test case to prove that a simple writer and reader roundtrip can work successfully without Hadoop dependency?

We do not yet have any serious ways for users to use the API without Hadoop dependencies. The added parameters to the TestReadWrite fixture make sure the read/write API should still function when using the ParquetConfiguration interface.

File: pom.xml
Comment (wgtmac) [inline pom.xml:550]: Just curious: can we remove this in a future PR? Only this PR introduces the breaking change here.

**Review by wgtmac — APPROVED** [review]
Thanks for the detail explanation! This LGTM. Probably you can also create JIRA issues in advance for all planned work items.

I would listen to feedback from more maintainers to proceed. @gszadovszky @shangxinli @ggershinsky @Fokko

File: pom.xml
Comment (amousavigourabi) [reply pom.xml:550]: I'm not that well versed in how japicmp works exactly, but I reckon this can be removed after the next minor release.

**Comment by wgtmac** [comment]
@shangxinli @gszadovszky Do you have any comment? If not, I plan to merge this next week. Thanks!

@amousavigourabi Could you please resolve the conflicts? Thank you!

**Comment by gszadovszky** [comment]
@wgtmac, @amousavigourabi, sorry, but I won't have the time to properly review this PR in the upcoming days.
The concept looks good to me so I am giving a +0 for this patch.

File: parquet-common/src/main/java/org/apache/parquet/conf/ParquetConfiguration.java
Comment (ConeyLiu) [inline parquet-common/src/main/java/org/apache/parquet/conf/ParquetConfiguration.java:27]: Maybe add some docs for the methods?

File: parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java:22]: unused import

File: parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java:237]: Seems this constructor is not necessary.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectZstd.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectZstd.java:150]: It should be safe to modify the existing method since it's a private method.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java:216]: Same here, we could change `getUnboundRecordFilterInstance` directly as well.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetWriter.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetWriter.java:296]: indent, it would be better to keep the same as the above method.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/api/ReadSupport.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/api/ReadSupport.java:89]: Do we still need to add a `deprecated` method?

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/api/ReadSupport.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/api/ReadSupport.java:138]: The error message should be more meaningful.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/api/WriteSupport.java
Comment (ConeyLiu) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/api/WriteSupport.java:115]: Same here

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroReadSupport.java
Comment (ConeyLiu) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroReadSupport.java:177]: Same here, we could update the existing private method directly.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java
Comment (ConeyLiu) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java:419]: same here

File: parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftRecordConverter.java
Comment (ConeyLiu) [inline parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftRecordConverter.java:864]: blank line

**Comment by ConeyLiu** [comment]
@amousavigourabi we have added so many public methods to keep the backward compatibility. So which one is preferred (I think it should be the`ParquetConfiguration`)? Should we deprecate the old one?

File: parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java
Comment (amousavigourabi) [reply parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java:237]: As this is public, removing the implicit no-args constructor _could_ potentially cause problems for users.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/api/ReadSupport.java
Comment (amousavigourabi) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/api/ReadSupport.java:89]: This PR is focussed on transitioning from `Configuration` to the `ParquetConfiguration` interface. This included some calls to deprecated methods which I could not very quickly transition away from. I would consider this out-of-scope for this PR.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/api/ReadSupport.java
Comment (amousavigourabi) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/api/ReadSupport.java:138]: I follow the example set by `ReadSupport#init(Configuration, Map, MessageType)`. As this error will not occur unless you are implementing your own `ReadSupport` class, I am not sure whether there needs to be that much more information in the exception. I'll add a reference to the `ReadSupport` class though.

**Comment by amousavigourabi** [comment]
> @amousavigourabi we have added so many public methods to keep the backward compatibility. So which one is preferred (I think it should be the`ParquetConfiguration`)? Should we deprecate the old one?

the `ParquetConfiguration` methods are indeed preferred. I would be in favour of deprecating the Hadoop `Configuration` methods. However, I do think deprecating these methods might warrant a separate discussion on the mailing list (as their deprecation and subsequent removal would go further than what we have previously discussed, it would be a step in the direction of eliminating the need for `hadoop-client-api` altogether instead of just `hadoop-client-runtime`) and does not necessarily need to be a part of this PR.

**Comment by wgtmac** [comment]
Thank you for the review @ConeyLiu. This PR is one of the efforts to use the parquet library without the Hadoop dependency. At the moment, we do not want to deprecate Hadoop dependency yet.

File: parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java
Comment (ConeyLiu) [reply parquet-hadoop/src/main/java/org/apache/parquet/ParquetReadOptions.java:236]: I see the usage, that makes sense.

**Review by ConeyLiu — APPROVED** [review]
Thanks, @amousavigourabi for the response and your efforts. LGTM.

> This PR is one of the efforts to use the parquet library without the Hadoop dependency. At the moment, we do not want to deprecate Hadoop dependency yet.

@wgtmac This is a good start. It would be nice to have a best practice guide doc to help users understand the different configuration usage.

**Comment by wgtmac** [comment]
I totally agree with you. @ConeyLiu 
We might need to consolidate different docs for configuration like this (https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/README.md) into a better space.

File: parquet-common/src/main/java/org/apache/parquet/conf/ParquetConfiguration.java
Comment (shangxinli) [inline parquet-common/src/main/java/org/apache/parquet/conf/ParquetConfiguration.java:35]: setString()?

Comment (amousavigourabi) [reply parquet-common/src/main/java/org/apache/parquet/conf/ParquetConfiguration.java:35]: I just went with the way Hadoop's `Configuration` does it for this and other methods in the interface, in order to not have to change references to conf.set() everywhere.

**Review by wgtmac — APPROVED** [review]

**Comment by wgtmac** [comment]
I will merge this if there is no more comment by the end of this week. Thanks @amousavigourabi for working on this!

**Comment by ConeyLiu** [comment]
This broke the CI checking.

File: parquet-common/src/main/java/org/apache/parquet/util/Reflection.java
Comment (ConeyLiu) [inline parquet-common/src/main/java/org/apache/parquet/util/Reflection.java:35]: Not sure why this breaks the API checking. However, this class seems redundant since there is not any usage.

Comment (wgtmac) [reply parquet-common/src/main/java/org/apache/parquet/util/Reflection.java:35]: Hmm, this is strange. Should we exclude this class in the japicmp config? @amousavigourabi

Comment (amousavigourabi) [reply parquet-common/src/main/java/org/apache/parquet/util/Reflection.java:35]: As there is no current usage, we can get away with deleting this class for now. My main concern with the CI failure is the incompatibility at `org.apache.parquet.conf.ParquetConfiguration.getClass(java.lang.String,java.lang.Class,java.lang.Class):CLASS_GENERIC_TEMPLATE_CHANGED` japicmp complains about. Especially as this did not come up before and the interface is completely new I have no idea why this is happening. Neither of the matters japicmp flags should be an issue AFAIK, so that might be something to investigate in our configuration?

Comment (ConeyLiu) [reply parquet-common/src/main/java/org/apache/parquet/util/Reflection.java:35]: I think this should be a bug of japicmp. Because both of the methods are generic template methods. And it should be safe to suppress the checking.

Comment (amousavigourabi) [reply parquet-common/src/main/java/org/apache/parquet/util/Reflection.java:35]: I'll open a new PR shortly, deleting the unused stuff and suppressing the warnings.

Comment (amousavigourabi) [reply parquet-common/src/main/java/org/apache/parquet/util/Reflection.java:35]: It has also started failing at other completely new methods in the Parquet Hadoop module: `org.apache.parquet.conf.HadoopParquetConfiguration.getClass(java.lang.String,java.lang.Class,java.lang.Class):CLASS_GENERIC_TEMPLATE_CHANGED,org.apache.parquet.hadoop.util.SerializationUtil.readObjectFromConfAsBase64(java.lang.String,org.apache.parquet.conf.ParquetConfiguration):CLASS_GENERIC_TEMPLATE_CHANGED`. Might be something to report to the japicmp maintainer?

Comment (ConeyLiu) [reply parquet-common/src/main/java/org/apache/parquet/util/Reflection.java:35]: I tested 0.16.0 has no problem. A little strange.

Comment (wgtmac) [reply parquet-common/src/main/java/org/apache/parquet/util/Reflection.java:35]: Probably the current 0.18.1 is more aggressive than 0.16.0

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java
Comment (Fokko) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:51]: This causes the Iceberg build to break when bumping to 1.14.0-SNAPSHOT:

```
> Task :iceberg-parquet:compileJava FAILED
/Users/fokkodriesprong/Desktop/iceberg/parquet/src/main/java/org/apache/iceberg/parquet/ParquetCodecFactory.java:60: error: cannot find symbol
        codecClass = configuration.getClassLoader().loadClass(codecClassName);
                                  ^
  symbol:   method getClassLoader()
  location: variable configuration of type ParquetConfiguration
/Users/fokkodriesprong/Desktop/iceberg/parquet/src/main/java/org/apache/iceberg/parquet/ParquetCodecFactory.java:62: error: incompatible types: ParquetConfiguration cannot be converted to Configuration
      codec = (CompressionCodec) ReflectionUtils.newInstance(codecClass, configuration);
```

Comment (wgtmac) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:51]: Does it make sense to fix this on the iceberg side?

Comment (wgtmac) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:51]: Anyway this is a breaking change. We have to fix it before the release.

Comment (Fokko) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/CodecFactory.java:51]: For Iceberg this is not an issue, since we would remove this class with the 1.14 release (that contains https://github.com/apache/parquet-mr/pull/1134): 
https://github.com/apache/iceberg/blob/866021d7d34f274349ce7de1f29d113395e7f28c/parquet/src/main/java/org/apache/iceberg/parquet/ParquetCodecFactory.java#L28-L32

In Iceberg we shade Parquet, so this will not interfere with other libraries that use a different version of Parquet. I'm curious what if we also have this issue on the Spark side (cc @vinooganesh).

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java
Comment (shangxinli) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java:204]: @amousavigourabi  Can you share more details on why this piece of code is removed?

Comment (amousavigourabi) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java:204]: looking back, i haven't a clue

**Pull request merged**
